File: dsa/array/searching/linear_search.js
/**
 * Algorithm: Linear Search
 * 
 * Time Complexity:
 *  -Best Case: O(1) When the element is found at the beginning
 *  -Worst Case: O(n) When the element is found at the end or not present
 *  -Average Case: O(n)
 * 
 * Space Complexity: O(1) (in-place i.e no extra space)
 * 
 * Description:
 *   Linear Search is the simplest searching algorithm
 *  It checks each element of the array one by one until the target element is found
 *   or the array ends.
 */

/**
 * Performs a linear search on an array to find the target element.
 * @param {Array} arr - The array to search in.
 * @param {*} target - The element to find.
 * @returns {number} The index of the target if found, else -1.
 */
function linearSearch(arr, target) {
    for (let i = 0; i < arr.length; i++) {
        if (arr[i]===target) {
            //here, === compares the value as well as the datatype
            return i; // Target found at index i
        }
    }
    return -1; 
}

//returning -1 indicates that the target is not found in the array

//Example 1: for integers
const arr1 = [10, 23, 45, 70, 11, 15];
const target1 = 70;
console.log(`Array: ${arr1}`);
console.log(`Searching for ${target1} --> Index Found:`, linearSearch(arr1, target1));

//Example 2: for strings
const arr2 = ["apple", "banana", "cherry", "date"];
const target2 = "kiwi";
console.log(`\nArray: ${arr2}`);
console.log(`Searching for "${target2}" --> Index Found:`, linearSearch(arr2, target2));

//Example 3:
const arr3 = [1, 3, 5, 7, 9];
const target3 = 5;
console.log(`\nArray: ${arr3}`);
console.log(`Searching for ${target3} --> Index Found:`, linearSearch(arr3, target3));

File: dsa_in_java/searching/LinearSearch.java
// File: LinearSearch.java
// Description: Implementation of Linear Search (Iterative & Recursive) in Java
// ---------------------------------------------------------------------------

public class LinearSearch {

    /**
     * Performs linear search iteratively.
     *
     * @param arr The array to search
     * @param target The value to find
     * @return The index of target if found, otherwise -1
     */
    public static int linearSearchIterative(int[] arr, int target) {
        for (int i = 0; i < arr.length; i++) {
            if (arr[i] == target)
                return i;
        }
        return -1;
    }

    /**
     * Performs linear search recursively.
     *
     * @param arr The array to search
     * @param target The value to find
     * @param index The current index
     * @return The index of target if found, otherwise -1
     */
    public static int linearSearchRecursive(int[] arr, int target, int index) {
        if (index >= arr.length)
            return -1;
        if (arr[index] == target)
            return index;
        return linearSearchRecursive(arr, target, index + 1);
    }

    // Main method for testing
    public static void main(String[] args) {
        int[] numbers = {10, 23, 45, 70, 11, 15};
        int target = 70;

        System.out.println("Array: [10, 23, 45, 70, 11, 15]");
        System.out.println("Searching for: " + target);

        int iterativeResult = linearSearchIterative(numbers, target);
        int recursiveResult = linearSearchRecursive(numbers, target, 0);

        System.out.println("\nIterative Linear Search Result: " + iterativeResult);
        System.out.println("Recursive Linear Search Result: " + recursiveResult);

        // Edge case: element not found
        int notFound = 99;
        System.out.println("\nSearching for element not in array (" + notFound + "):");
        System.out.println("Result: " + linearSearchIterative(numbers, notFound));
    }
}

/*
--------------------------------
Example Output:

Array: [10, 23, 45, 70, 11, 15]
Searching for: 70

Iterative Linear Search Result: 3
Recursive Linear Search Result: 3

Searching for element not in array (99):
Result: -1
--------------------------------

Time Complexity:
- Best Case: O(1)
- Average & Worst Case: O(n)
Space Complexity:
- Iterative: O(1)
- Recursive: O(n) (due to call stack)
*/
